[FX] Fixes for pickling GraphModule instances

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44006 [FX] Fixes for pickling GraphModule instances**

1. Only assign `forward` over from `gbls` when we're `exec`ing code. Previously this would dump things like `torch` on Module instances as attributes, which messed up pickling elsewhere
2. During deserialization, create a new local type and assign `forward` on that. `symbolic_trace` pulls the `forward` function off the type of the passed-in module rather than the instance, so without this fix there were situations where we would compile the wrong code

Differential Revision: [D23466542](https://our.internmc.facebook.com/intern/diff/D23466542)

Commit:
[FX] Only copy over forward() from exec

[ghstack-poisoned]

Commit:
Update on "[FX] Only copy over forward() from exec"


Differential Revision: [D23466542](https://our.internmc.facebook.com/intern/diff/D23466542)

[ghstack-poisoned]

zdevito commented on torch/fx/graph_module.py:
```diff
@@ -35,16 +35,21 @@ def deserialize_graphmodule(root : torch.nn.Module, src : str) -> torch.nn.Modul
+    # function off of the class, rather than the instance
+    class Dummy(torch.nn.Module):
+        def __init__(self, root, src):
+            super().__init__()
```
nit: its a bit confusing to do class wide modifications in an instance. This would perhaps be clearer:
```
class Dummy(...):
   def __init__(self, root):
      self.root = root

Dummy.forward = generate_forward(src)
```

Also dummy is not a descriptive name, `CodeOnlyModule`? Not great either but it should describe what the object contains.

zdevito commented on torch/fx/graph_module.py:
```diff
@@ -82,8 +87,7 @@ def forward(self, {', '.join(free_variables)}):
         cls = type(self)
-        for k, v in gbls.items():
-            setattr(cls, k, v)
+        setattr(cls, 'forward', gbls['forward'])
```
exec with source and globals extraction now occurs in two places, should be a function like `generate_forward` that just returns the function implementation given the source and handles the globals creation.

zdevito reviewed (approved):
Looks good, minor nits.

Commit:
Update on "[FX] Fixes for pickling GraphModule instances"


1. Only assign `forward` over from `glbs` when we're `exec`ing code. Previously this would dump things like `torch` on Module instances as attributes, which messed up pickling elsewhere
2. During deserialization, create a new local type and assign `forward` on that. `symbolic_trace` pulls the `forward` function off the type of the passed-in module rather than the instance, so without this fix there were situations where we would compile the wrong code

Differential Revision: [D23466542](https://our.internmc.facebook.com/intern/diff/D23466542)

[ghstack-poisoned]
